### ISSUE TITLE
crash caused by BUS_ADRALN on armeabi-v7a

### ISSUE THREAD

**Issue by jojo59516** [LuaJIT/LuaJIT#608]
Hi,

We use the excellent LuaJIT in our Unity project. Recently we are trying to cross-compile `luajit.a` by Clang instead of GCC on Windows (using Cygwin), and encounter the crash caused by `signal 7 (SIGBUS), code 1 (BUS_ADRALN)` on Android device with `armeabi-v7a` CPU. 

The crash occurs at the line
```C
/* src/lj_trace.c:438 */
memset(&J->cur, 0, sizeof(GCtrace));
```
in `src/lj_trace.c`, where `&J->cur` is `0xc5b3d3f8`, not aligned to16 bytes.

Version: 2.1. Git hash: 570e758ca7dd14f93efdd43d68cf8979c1d7f984

**Comment by jojo59516** [comment]
And here's our `build.sh`: 
```bash
#!/bin/bash

NDKDIR=C:/Users/Administrator/AppData/Local/Android/Sdk/ndk/21.3.6528147
NDKBIN=$NDKDIR/toolchains/llvm/prebuilt/windows-x86_64/bin

NDKCROSS=$NDKBIN/arm-linux-androideabi-
NDKCC=$NDKBIN/armv7a-linux-androideabi21-clang

make clean
make HOST_CC="gcc -m32" CROSS=$NDKCROSS \
     STATIC_CC=$NDKCC DYNAMIC_CC="$NDKCC -fPIC" \
     TARGET_LD=$NDKCC TARGET_SYS=Linux \
	 TARGET_FLAGS="-isysroot $NDKDIR/platforms/android-21/arch-arm -mfloat-abi=softfp" \
	 XCFLAGS="-DLUAJIT_ENABLE_LUA52COMPAT"
dst=../lib/android/armeabi-v7a/
mkdir -p $dst && cp src/libluajit.a $dst
make clean
```

**Comment by jojo59516** [comment]
And here's the errors in log:
```
Version '2019.4.5f1 (81610f64359c)', Build type 'Development', Scripting Backend 'mono', CPU 'armeabi-v7a'
Build fingerprint: 'samsung/dreamqltezc/dreamqltechn:9/PPR1.180610.011/G9500ZCU5DTF3:user/release-keys'
Revision: '12'
ABI: 'arm'
Timestamp: 2020-08-04 19:54:01+0800
pid: 31530, tid: 31577, name: UnityMain  >>> com.pwrd.ark.nosdk <<<
uid: 10832
signal 7 (SIGBUS), code 1 (BUS_ADRALN), fault addr 0xc5b3d3f8
    r0  c5b3d3f8  r1  00000000  r2  00000000  r3  00000000
    r4  c5b3d3b8  r5  00008000  r6  00000000  r7  00000001
    r8  c5b3d1c8  r9  c5b3d1c8  r10 ffffffa0  r11 d3d4ddd8
    ip  00000000  sp  d3d4dcb8  lr  c11814a8  pc  c11814e8

backtrace:
      #00 pc 0010a4e8  /data/app/com.pwrd.ark.nosdk-FdagC_fXCXR6pY6KMF2R4w==/lib/arm/libruntime_client.so (trace_start+320) (BuildId: b73b6d7ef00d40f5bb357ce4f0bb94c1f1c0cc10)
      #01 pc 001092c8  /data/app/com.pwrd.ark.nosdk-FdagC_fXCXR6pY6KMF2R4w==/lib/arm/libruntime_client.so (trace_state+140) (BuildId: b73b6d7ef00d40f5bb357ce4f0bb94c1f1c0cc10)
      #02 pc 001140b4  /data/app/com.pwrd.ark.nosdk-FdagC_fXCXR6pY6KMF2R4w==/lib/arm/libruntime_client.so (lj_vm_cpcall+76) (BuildId: b73b6d7ef00d40f5bb357ce4f0bb94c1f1c0cc10)
      #03 pc 00109ce0  /data/app/com.pwrd.ark.nosdk-FdagC_fXCXR6pY6KMF2R4w==/lib/arm/libruntime_client.so (lj_trace_hot+164) (BuildId: b73b6d7ef00d40f5bb357ce4f0bb94c1f1c0cc10)
      #04 pc 00115654  /data/app/com.pwrd.ark.nosdk-FdagC_fXCXR6pY6KMF2R4w==/lib/arm/libruntime_client.so (lj_vm_hotloop+40) (BuildId: b73b6d7ef00d40f5bb357ce4f0bb94c1f1c0cc10)
      #05 pc 000fd0e8  /data/app/com.pwrd.ark.nosdk-FdagC_fXCXR6pY6KMF2R4w==/lib/arm/libruntime_client.so (lua_pcall+92) (BuildId: b73b6d7ef00d40f5bb357ce4f0bb94c1f1c0cc10)
      #06 pc 0000a5f6  <anonymous:c5b86000>

managed backtrace:
      #00 (wrapper managed-to-native) LuaInterface.LuaDLL:lua_pcall (intptr,int,int,int)
      #01 LuaInterface.LuaDLL:luaL_dostring (intptr,string) <E:\ark_jenkins\workspace\build_apk\ark_client\client\Assets\ThirdParty\ToLua\Core\LuaDLL.cs:901>
      #02 LuaInterface.LuaStatePtr:LuaOpenJit () <E:\ark_jenkins\workspace\build_apk\ark_client\client\Assets\ThirdParty\ToLua\Core\LuaStatePtr.cs:83>
      #03 LuaInterface.LuaStatePtr:OpenToLuaLibs () <E:\ark_jenkins\workspace\build_apk\ark_client\client\Assets\ThirdParty\ToLua\Core\LuaStatePtr.cs:627>
      #04 LuaInterface.LuaState:.ctor () <E:\ark_jenkins\workspace\build_apk\ark_client\client\Assets\ThirdParty\ToLua\Core\LuaState.cs:111>
      #05 Framework.ScriptManager:Init (Framework.ScriptManager/InitType) <E:\ark_jenkins\workspace\build_apk\ark_client\client\Assets\Source\Framework\Script\ScriptManager.cs:115>
      #06 Client.Client/<_Init>d__8:MoveNext () <E:\ark_jenkins\workspace\build_apk\ark_client\client\Assets\Source\Framework\Client.cs:89>
      #07 UnityEngine.SetupCoroutine:InvokeMoveNext (System.Collections.IEnumerator,intptr) </Users/builduser/buildslave/unity/build/Runtime/Export/Scripting/Coroutines.cs:17>
      #08 (wrapper runtime-invoke) <Module>:runtime_invoke_void_object_intptr (object,intptr,intptr,intptr)
```
Here's the [full log](https://github.com/LuaJIT/LuaJIT/files/5025681/tombstone_00.log).

**Comment by MikePall** [comment]
This is an issue with the toolchain you're using to build LuaJIT  and not an issue with LuaJIT itself.

The `memset()` call is translated by the C compiler into either a library call or some inline code. For the latter, the C compiler has to consider the maximum alignment of the struct (which is 4 for GCtrace on a 32 bit platform) and the alignment restrictions of the target platform wrt. the instructions used in the inlined code.

If there's a mismatch, e.g. because Clang generates SIMD instructions to clear a 4-byte aligned struct, but the hardware requires 16-byte alignment for SIMD instructions, you'll get a crash.

In other words: if you built the toolchain yourself, you have not properly configured it. If you took an existing toolchain (e.g. from the NDK), then please complain to whoever published that toolchain (e.g. Google).

**Comment by dwing4g** [comment]
@MikePall 
jit_State is defined with "LJ_ALIGN(16)" on ARM. The compiler assumes this and emits 16-byte vector store instruction for that memset.
However, jit_State is defined in GG_State that is allocated by allocf in lua_newstate, and allocf does NOT guarantee how to align.
So the jit_State may not be aligned correctly in fact.

**Comment by MikePall** [comment]
@dwing4g Yes, you're right. I missed that the C compiler can infer the inner struct alignment.

I've committed a fix. Thanks!
